src/DocumentTemplate/_DocumentTemplate.py

```diff
@@ -103,83 +103,80 @@
 from DocumentTemplate.ustr import ustr
 
-if sys.version_info > (3, 0):
-    basestring = str
-    unicode = str
-    classtype = type
-else:
-    classtype = types.ClassType
+import DocumentTemplate as _dt
 
 _marker = object()
 
 
-def join_unicode(rendered):
+def join_unicode(rendered, encoding=None):
     """join a list of plain strings into a single plain string,
     a list of unicode strings into a single unicode strings,
     or a list containing a mix into a single unicode string with
-    the plain strings converted from latin-1
+    the plain strings converted from the given ``encoding``, or
+    Latin-1 as default fallback.
     """
     try:
         return ''.join(rendered)
-    except UnicodeError:
+    except (UnicodeError, TypeError):
         # A mix of unicode string and non-ascii plain strings.
-        # Fix up the list, treating normal strings as latin-1
+        # Fix up the list, treating normal strings as encoded in the
+        # passed-in ``encoding``, or Latin-1 as fallback.
+        if encoding is None:
+            encoding = _dt.DEFAULT_ENCODING
         rendered = list(rendered)
         for i in range(len(rendered)):
-            if isinstance(rendered[i], str):
-                rendered[i] = unicode(rendered[i], 'latin-1')
+            if isinstance(rendered[i], six.binary_type):
+                rendered[i] = rendered[i].decode(encoding)
         return u''.join(rendered)
 
 
-def render_blocks(blocks, md):
+def render_blocks(blocks, md, encoding=None):
     rendered = []
 
-    render_blocks_(blocks, rendered, md)
+    render_blocks_(blocks, rendered, md, encoding)
 
     l_ = len(rendered)
     if l_ == 0:
         return ''
     elif l_ == 1:
         return rendered[0]
-    return join_unicode(rendered)
+    return join_unicode(rendered, encoding=encoding)
 
 
-def render_blocks_(blocks, rendered, md):
+def render_blocks_(blocks, rendered, md, encoding):
     for block in blocks:
         append = True
 
         if (isinstance(block, tuple) and
                 len(block) > 1 and
-                isinstance(block[0], basestring)):
+                isinstance(block[0], six.string_types)):
 
             first_char = block[0][0]
             if first_char == 'v':  # var
                 t = block[1]
-                if isinstance(t, str):
+                if isinstance(t, six.string_types):
                     t = md[t]
                 else:
                     t = t(md)
 
                 skip_html_quote = 0
-                if not isinstance(t, basestring):
+                if not isinstance(t, (six.string_types, six.binary_type)):
                     # This might be a TaintedString object
                     untaintmethod = getattr(t, '__untaint__', None)
                     if untaintmethod is not None:
                         # Quote it
                         t = untaintmethod()
                         skip_html_quote = 1
 
-                if not isinstance(t, basestring):
```

Comment:
How did this work with all the `unicode` and `basestring` names?

Reply:
Ah, I see, we were defining them at the beginning of the module.

Reply:
There's many places where code wants to decide "is this a string" as opposed to, say, a callable. The work that's been done to make all that Python 3 compatible was really haphazard, inconsistent and incomplete. I've spent hours stepping through the debugger on this, and I am dure there will be more edge cases found in the future.